internal/telemetry/tracker.go

```diff
@@ -96,8 +119,369 @@ func (t *Tracker) trackClusterDeploymentHeartbeat(ctx context.Context) error {
+func countNonSystemPodsOnNodes(ctx context.Context, k8sClient client.Client) (map[string]map[string]int, error) {
+	// Get all pods in the cluster -- these will be filtered to just those assigned to the node
+	pods := &corev1.PodList{}
+	if err := k8sClient.List(ctx, pods); err != nil {
+		return nil, err
+	}
+
+	nodePodMetrics := make(map[string]map[string]int)
+
+	// Count pods
+	for _, pod := range pods.Items {
+		// Skip system namespace pods using rubric
+		if systemNamespaces[pod.Namespace] {
+			continue
+		}
+
+		// Skip pods not assigned to a node
+		if pod.Spec.NodeName == "" {
+			continue
+		}
+
+		// Initialize node metrics if not exists
+		if nodePodMetrics[pod.Spec.NodeName] == nil {
+			nodePodMetrics[pod.Spec.NodeName] = map[string]int{
+				"nodePodCount":           0,
+				"nodePodRunningCount":    0,
+				"nodePodGpuRequestCount": 0,
+				"nodePodGpuAllocCount":   0,
+				"containerCount":         0,
+			}
+		}
+
+		// Count total non-system pods and containers
+		nodePodMetrics[pod.Spec.NodeName]["nodePodCount"]++
+		nodePodMetrics[pod.Spec.NodeName]["containerCount"] += len(pod.Spec.Containers) + len(pod.Spec.InitContainers)
+
+		// Running pods
+		if pod.Status.Phase == corev1.PodRunning {
+			nodePodMetrics[pod.Spec.NodeName]["nodePodRunningCount"]++
+		}
+
+		// Pods requesting GPU -- need to check resources associated with Pod containeres
+		podRequestGPU := false
+		podAllocGPU := false
+		for _, container := range pod.Spec.Containers {
+			if container.Resources.Requests != nil {
+				for resourceName := range container.Resources.Requests {
+					if isGPUResource(resourceName) {
+						podRequestGPU = true
+						break
+					}
+				}
+			}
+			if container.Resources.Limits != nil {
+				for resourceName := range container.Resources.Limits {
+					if isGPUResource(resourceName) {
+						podAllocGPU = true
+						break
+					}
+				}
+			}
+		}
+
+		if podRequestGPU {
+			nodePodMetrics[pod.Spec.NodeName]["nodePodGpuRequestCount"]++
+		}
+		if podAllocGPU {
+			nodePodMetrics[pod.Spec.NodeName]["nodePodGpuAllocCount"]++
+		}
+	}
+
+	return nodePodMetrics, nil
+}
+
+// Helper function to build node-level metrics
+// Return a map indexed by node name, plus cluster-level aggregates
+func buildNodeLevelMetrics(ctx context.Context, k8sClient client.Client) ([]map[string]any, map[string]any, error) {
+	// Get all nodes in the cluster
+	nodes := &corev1.NodeList{}
+	if err := k8sClient.List(ctx, nodes); err != nil {
+		return nil, nil, err
+	}
+
+	clusterMetrics := make(map[string]any)
+	nodeMetrics := make([]map[string]any, 0)
+	controllerNodes := 0
+	workerNodes := 0
+	totalPods := 0
+	totalContainers := 0
+	totalPodsRequestGPU := 0
+	totalPodsAllocGPU := 0
+
+	// Count non-system pods per node
+	nodePodCounts, err := countNonSystemPodsOnNodes(ctx, k8sClient)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Iterate through nodes and build a node-level metrics map
+	for i, node := range nodes.Items {
```

Comment:
That will work only on a small number of nodes; otherwise, either OOM or overusage of resources to collect metrics once in a tick. Such a core's objects must be collected and aggregated in batches